cross-dom-comm/README.md

```diff
@@ -0,0 +1,237 @@
+[![Twitter Follow](https://img.shields.io/twitter/follow/optimismPBC.svg?label=optimismPBC&style=social)](https://twitter.com/optimismPBC)
+
+This tutorial teaches you how to write a contract on one layer (either L1 or L2) that sends a message to call a function on a contract on the other layer. 
+[You can read more details about this process here](https://community.optimism.io/docs/developers/bridge/messaging/).
+
+## Seeing it in action
+
+To show how this works we installed [a slightly modified version of HardHat's `Greeter.sol`](contracts/Greeter.sol) on both L1 Kovan and Optimistic Kovan
+
+
+| Network | Greeter address  |
+| ------- | ---------------- |
+| Kovan (L1) | [0x11fB328D5Bd8E27917535b6d40b881d35BC39Be0](https://kovan.etherscan.io/address/0x11fB328D5Bd8E27917535b6d40b881d35BC39Be0) |
+| Optimistic Kovan (L2) | [0xD4c204223d6F1Dfad0b7a0b05BB0bCaB6665e0c9](https://kovan-optimistic.etherscan.io/address/0xD4c204223d6F1Dfad0b7a0b05BB0bCaB6665e0c9) |
+
+
+### Setup
+
+This setup assumes you already have [Node.js](https://nodejs.org/en/) and [yarn](https://classic.yarnpkg.com/) installed on your system. 
+
+1. Edit `hardhat.config.js`:
+   1. Set `mnemonic` to point to an account that has ETH on the Kovan test network. 
+   1. Set `module.exports.networks.kovan` to point to a URL that accesses the Kovan test network.
+
+1. Install the necessary packages.
+
+   ```sh
+   yarn
+   ```
+
+### L1 message to L2
+
+1. [Browse to the Greeter contract on Etherscan](https://kovan-optimistic.etherscan.io/address/0xD4c204223d6F1Dfad0b7a0b05BB0bCaB6665e0c9#readContract) and click **greet** to see the greeting.
+
+1. Connect the Hardhat console to Kovan (L1):
+
+   ```sh
+   yarn hardhat console --network kovan
+   ```
+
+1. Deploy and call the `ControlL2Greeter` contract.
+
+   ```js
+   Controller = await ethers.getContractFactory("ControlL2Greeter")
+   controller = await Controller.deploy()
+   tx = await controller.setGreeting("Shalom")
+   rcpt = await tx.wait()
+   ```
+
+1. Make a note of the address of `ControlL2Greeter`.
+
+   ```js
+   controller.address
+   ```
+
+1. [Browse to the Greeter contract on Etherscan](https://kovan-optimistic.etherscan.io/address/0xD4c204223d6F1Dfad0b7a0b05BB0bCaB6665e0c9#readContract) and click **greet** to see the new greeting.
+
+
+
+### L2 message to L1
+
+
+1. [Browse to the Greeter contract on Etherscan](https://kovan.etherscan.io/address/0x11fB328D5Bd8E27917535b6d40b881d35BC39Be0#readContract) and click **greet** to see the greeting.
+
+1. Connect the Hardhat console to Optimistic Kovan (L2):
+
+   ```sh
+   yarn hardhat console --network kovan
+   ```
+
+1. Deploy and call the `ControlL1Greeter` contract.
+
+   ```js
+   Controller = await ethers.getContractFactory("ControlL1Greeter")
```

Comment:
It feels a bit clunky to have two different contracts (the Greeter and the Controller). Can we just merge into a single contract instead?

Reply:
I think keeping it to one contract will make this much simpler. I also like the idea of having the contracts deployed to a live network, but I also think we should give people instructions for running the tutorial locally.

Reply:
I don't think I understand you. `Greeter` and `Controller` live on different layers. The whole point is to have a contract on one layer control a contract on another.

Do you think it will be simpler if we have one contract that is both a Greeter and a Controller for the Greeter on the other layer? How is one contract doing two things simpler than two contracts that only do one thing each?

Or do you mean that it would be simpler to have one Controller contract that we can deploy on both layer 1 and layer 2, and that identifies its layer programmatically?

I'll add how to run the tutorial locally.

Reply:
I do it like this is that Greeter represents a previously existing contract that you need to interface with. Controller is the contract on the other layer that the reader is expected to write.